renderdoc/driver/d3d12/d3d12_pixelhistory.cpp

```diff
@@ -0,0 +1,401 @@
+  else
+    uavIndex = 4;
+
+  struct CopyPixelShaderInput
+  {
+    Vec4u src_coord;    // x, y, mip/sample, slice
+
+    uint32_t dst_slot;
+    uint32_t copy_depth;
+    uint32_t copy_stencil;
+
+    uint32_t multisampled;
+    uint32_t is_float;
+    uint32_t is_uint;
+    uint32_t is_int;
+  } inputData;
+
+  inputData.src_coord = {p.x, p.y, p.multisampled ? p.sample : p.mip, 0};
+  inputData.multisampled = p.multisampled;
+  inputData.is_float = floatTex;
+  inputData.is_uint = uintTex;
+  inputData.is_int = intTex;
+
+  inputData.dst_slot = (uint32_t)(offset / sizeof(float));
+  inputData.copy_depth = p.depthcopy && p.planeSlice == 0;
+  inputData.copy_stencil = p.depthcopy && p.planeSlice == 1;
+
+  D3D12_CPU_DESCRIPTOR_HANDLE srv =
+      m_pDevice->GetDebugManager()->GetCPUHandle(FIRST_PIXELHISTORY_SRV);
+  D3D12_CPU_DESCRIPTOR_HANDLE uav =
+      m_pDevice->GetDebugManager()->GetCPUHandle(FIRST_PIXELHISTORY_UAV);
+
+  // The shader has 10 SRVs which fits all scenarios we need, but we cannot trivially set
+  // all SRVs as a single root sig table param because the SRV needed for a given index can
+  // change, if we are reading a scratch buffer and then later the real resource being inspected.
+  // So instead, specify all 10 SRVs as root params and swpa out the index with a different SRV
+  // ptr as needed.
+  const UINT numPixelHistorySrvs = LAST_PIXELHISTORY_SRV - FIRST_PIXELHISTORY_SRV;
+  UINT readSrvIndex = srvIndex;
+  if(p.scratchBuffer)
+  {
+    if(p.depthcopy)
+    {
+      // Either depth or stencil, need a separate SRV for each
+      readSrvIndex = numPixelHistorySrvs + 1 + p.planeSlice;
+    }
+    else
+    {
+      // Color, should only need one SRV for this
+      readSrvIndex = numPixelHistorySrvs;
+    }
+  }
+  D3D12_CPU_DESCRIPTOR_HANDLE readSrv = srv;
+  readSrv.ptr += readSrvIndex * sizeof(D3D12Descriptor);
+
+  D3D12_SHADER_RESOURCE_VIEW_DESC srvDesc = {};
+  srvDesc.ViewDimension =
+      p.multisampled ? D3D12_SRV_DIMENSION_TEXTURE2DMS : D3D12_SRV_DIMENSION_TEXTURE2D;
+  srvDesc.Format = p.srcImageFormat;
+  srvDesc.Shader4ComponentMapping = D3D12_DEFAULT_SHADER_4_COMPONENT_MAPPING;
+  if(!p.multisampled)
+  {
+    srvDesc.Texture2D.MipLevels = 1;
+    srvDesc.Texture2D.PlaneSlice = p.planeSlice;
+  }
+  m_pDevice->CreateShaderResourceView(p.srcImage, &srvDesc, readSrv);
+
+  m_pDevice->GetDebugManager()->SetDescriptorHeaps(state.heaps, true, false);
+  state.compute.sigelems = {
+      D3D12RenderState::SignatureElement(
+          eRootCBV, m_pDevice->GetDebugManager()->UploadConstants(&inputData, sizeof(inputData))),
+      D3D12RenderState::SignatureElement(eRootTable, uav),
+  };
+
+  for(UINT i = 0; i < numPixelHistorySrvs; ++i)
+  {
+    if(i == srvIndex)
+      state.compute.sigelems.push_back(D3D12RenderState::SignatureElement(eRootTable, readSrv));
+    else
+      state.compute.sigelems.push_back(D3D12RenderState::SignatureElement(eRootTable, srv));
+
+    srv.ptr += sizeof(D3D12Descriptor);
+  }
```

Comment:
I don't fully understand the need for separate signature elements here or how this is intended to work. As I see it right now, you're creating the correct SRV in the corresponding correct fixed heap offset and binding that, then binding the depth SRV to all other slots. This means that the unused slots might be uninitialised, or else otherwise the wrong type. I don't remember what D3D12 says about the validity of this but I think it does allow unaccessed descriptors to be garbage so that part is fine.

It seems like the issue comes from the scratch SRVs as they're not contiguous in the same way, but I'm missing the context on why these need to be separate in the first place rather than using the same range?

Reply:
Null descriptors being bound are fine with D3D12, but what you can't do is overwrite a SRV in your heap within a single command list execution timeline. Sometimes we need to copy MSAA depth/stencil to get per-fragment output or post-draw depth, and other times we need to copy a scratch stencil (to get fragment count, etc.). Those are separate resources, so they can't occupy the same heap slot. To be able to limit the number of GPU syncs that pixel history does (there's several already), adding a separate heap slot for the SRV was the cleanest way to support all cases.

And since if you have a tableParam with many elements, they need to be contiguous in the heap, this required specifying each SRV separately, to account for one of them potentially being the non-contiguous scratch SRV.

I'll add a comment to explain this, at the root sig definition site.

Reply:
We talked about this further, and I've switched it to having a separate 10 SRV heap range for scratch copies, which made the code slightly simpler.